### PR TITLE
[ADD] Module web_calendar_year 8.0

### PR DESCRIPTION
**Pull request by astirpe**
Module Year calendar view.
Extends the default calendar widget with a year view.

### CODE REVIEW

**Comment by pedrobaeza** [comment]
Hi, Andrea, thanks for the contribution!

I'm reviewing it.

File: web_calendar_year/README.rst [hunk 1]
```diff
@@ -0,0 +1,3 @@
+Year calendar view
```
Comment (pedrobaeza) [inline web_calendar_year/README.rst]: Please use full OCA's README template: https://raw.githubusercontent.com/OCA/maintainer-tools/master/template/module/README.rst

File: web_calendar_year/__init__.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
+# -*- coding: utf-8 -*-
```
Comment (pedrobaeza) [inline web_calendar_year/__init__.py:1]: Use short headers: https://github.com/OCA/maintainer-tools/blob/master/template/module/__init__.py

File: web_calendar_year/__openerp__.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+{
+    'name': 'Web Calendar Year',
+    'images': ['static/description/main_screenshot.png'],
+    'summary': """Year view on Odoo calendar""",
+    'author': 'ONESTEiN BV,Odoo Community Association (OCA)',
+    'license': 'AGPL-3',
+    'website': 'http://www.onestein.eu',
+    'category': 'Extra Tools',
+    'version': '1.0',
```
Comment (pedrobaeza) [inline web_calendar_year/__openerp__.py]: 8.0.1.0.0

File: web_calendar_year/__openerp__.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+##############################################################################
+
+{
+    'name': 'Web Calendar Year',
+    'images': ['static/description/main_screenshot.png'],
+    'summary': """Year view on Odoo calendar""",
+    'author': 'ONESTEiN BV,Odoo Community Association (OCA)',
+    'license': 'AGPL-3',
+    'website': 'http://www.onestein.eu',
+    'category': 'Extra Tools',
```
Comment (pedrobaeza) [inline web_calendar_year/__openerp__.py]: Use category 'Web'

File: web_calendar_year/doc/index.rst [hunk 5]
```diff
@@ -0,0 +1,9 @@
+This module adds a year view to all Odoo calendar views for all users.
```
Comment (pedrobaeza) [inline web_calendar_year/doc/index.rst]: Remove this file and company references

**Comment by pedrobaeza** [comment]
2 more things:
- Remove Onenstein logo from the image.
- A more descriptive icon than the generic calendar one would be better.

File: web_calendar_year/__openerp__.py [hunk 6]
```diff
@@ -0,0 +1,22 @@
+# -*- coding: utf-8 -*-
+# © 2015 ONESTEiN BV (<http://www.onestein.eu>)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+{
+    'name': 'Web Calendar Year',
+    'images': ['static/description/main_screenshot.png'],
+    'summary': """Year view on Odoo calendar""",
```
Comment (hbrunn) [inline web_calendar_year/__openerp__.py]: no need for a multiline string here

File: web_calendar_year/__openerp__.py [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+    'category': 'Web',
+    'version': '8.0.1.0.0',
+    'depends': ['web_calendar'],
+    'data': [
+        'views/web_calendar_year.xml',
+    ],
+    'demo': [],
+    'installable': True,
+    'auto_install': False,
+    'application': False,
```
Comment (hbrunn) [inline web_calendar_year/__openerp__.py]: remove empty keys and keys with default value

File: web_calendar_year/static/src/js/web_calendar_year.js [hunk 8]
```diff
@@ -0,0 +1,39 @@
+openerp.web_calendar_year = function(instance) {
+
+    instance.web_calendar.CalendarView = instance.web_calendar.CalendarView.extend({
```
Comment (hbrunn) [inline web_calendar_year/static/src/js/web_calendar_year.js]: dont do this, this changes the reference of the object. Use `instance.web_calendar.CalendarView.include`

File: web_calendar_year/static/src/js/web_calendar_year.js [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+            options.header.right = "year,month,agendaWeek,agendaDay";
+			options.eventAfterAllRender = function() {
+				console.log("eventAfterAllRender");
+				var tableBody = $('.oe_view_manager_body');
+				tableBody.find('.fake-thead').remove();
+				tableBody.scroll();
+			}
+            return options;
+        },
+		init_calendar: function() {
```
Comment (hbrunn) [inline web_calendar_year/static/src/js/web_calendar_year.js]: unindent

File: web_calendar_year/static/src/js/web_calendar_year.js [hunk 10]
```diff
@@ -0,0 +1,39 @@
+openerp.web_calendar_year = function(instance) {
+
+    instance.web_calendar.CalendarView = instance.web_calendar.CalendarView.extend({
+        get_fc_init_options: function() {
+            var options = this._super();
+            options.buttonText.year = instance.web._t("Year");
+            options.header.right = "year,month,agendaWeek,agendaDay";
+			options.eventAfterAllRender = function() {
```
Comment (hbrunn) [inline web_calendar_year/static/src/js/web_calendar_year.js]: unindent

File: web_calendar_year/static/src/js/web_calendar_year.js [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+				console.log("eventAfterAllRender");
```
Comment (hbrunn) [inline web_calendar_year/static/src/js/web_calendar_year.js]: remove debug print

File: web_calendar_year/views/web_calendar_year.xml [hunk 12]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="utf-8"?>
+<openerp>
+    <data>
+        <template id="assets_backend" name="web_calendar_year assets" inherit_id="web.assets_backend">
+            <xpath expr="." position="inside">
+                <script type="text/javascript" src="/web_calendar_year/static/src/js/fullcalendar.js"></script>
```
Comment (hbrunn) [inline web_calendar_year/views/web_calendar_year.xml]: libraries belong in `/static/lib`

**Comment by hbrunn** [comment]
:+1:

**Comment by pedrobaeza** [comment]
We are introducing here code from Adam Shaw (the JS lib), so we need him to sign the CLA also.

**Comment by hbrunn** [comment]
@pedrobaeza is there any OCA-decision about that? I don't think it's feasible to have library authors sign our CLA, this effectively would mean we can't use existing libraries, that would be horrible

**Comment by hbrunn** [comment]
PS: As the libraries have their own copyright, I think it's clear that the OCA can't have any claims on them

**Comment by astirpe** [comment]
That js library (FullCalendar) is a modified copy of the js library belonging to the standard Odoo Module _web_calendar_

**Comment by hbrunn** [comment]
oh, I was not aware of that. Isn't there a possibility to use Odoo's version and monkey patch whatever you need? That would be a lot cleaner, because this way, you never know which version of the library you have in hand currently.

**Comment by astirpe** [comment]
@hbrunn indeed would be a lot cleaner! We will check asap for a better solution.

In the meanwhile, I would ask an example, in case someone is aware of any similar solution applied on a module (I mean import/refer js libs defined in a base module + extending or monkey-patching them). That would be very appreciated!

**Comment by hbrunn** [comment]
have a look at https://github.com/OCA/web/blob/8.0/web_ckeditor4/static/src/js/web_ckeditor4.js#L25 - this is not exactly your case, but should give you the idea

**Comment by astirpe** [comment]
@hbrunn thank you very much! Your example is very interesting. I'm still trying to figure out how to apply a similar solution for that FullCalendar library, I will be back on it as soon as possible ...

**Comment by hbrunn** [comment]
if you post the diff of what you needed to change, we might be able to help you

**Comment by astirpe** [comment]
Here is the [diff](https://github.com/astirpe/fullcalendar/commit/caafb6b053ab1061e1201aa4761e74883a8c907f) with the modifications applied to the original fullcalendar library.

As you can notice, the original code is entirely wrapped inside a kind of anonymous function:

```
(function($, undefined) {

    ...

})(jQuery);
```

an this is the main difficulty: extending or patching the original code contained in an anonymous function.

**Comment by hbrunn** [comment]
luckily, this code puts most of its references into `$.fullCalendar`. So where you say

``` javascript
fcViews.year = YearView;
function YearView(element, calendar)
...
```

you should be also able to say

``` javascript
jQuery.fullCalendar.views.year = function(element, calendar)
...
```

then I mostly see `console.logs` which we don't want so badly anyways and a little adaption of `BasicView` of which I'm not sure it it's necessary.

**Comment by pedrobaeza** [comment]
@astirpe, are you going to work in the cleaner implementation suggested by @hbrunn ?

**Comment by astirpe** [comment]
@pedrobaeza  not for the time being, sorry. Feel free to close the PR (if that was the intention)

**Comment by hbrunn** [comment]
let's set it to work in progress then, this way, it stays visible for whomever it concerns, but won't be shown in lists filtering for needs review. I also remove the CLA label as this is done too to my knowledge

**Comment by astirpe** [comment]
I will not continue the development of this module, so I'm closing this old PR.
The original module will remain available [here](https://github.com/onesteinbv/addons-onestein/tree/8.0/web_calendar_year).

**Pull request closed without merging**